cirq-core/cirq/testing/routing_devices.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Provides test devices that can validate circuits during a routing procedure."""
+
+from typing import Optional, TYPE_CHECKING
+import networkx as nx
+
+from cirq import devices
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class RoutingTestingDevice(devices.Device):
+    """Testing device to be used only for testing qubit connectivity in routing procedures."""
+
+    def __init__(self, metadata: devices.DeviceMetadata) -> None:
+        self._metadata = metadata
+
+    @property
+    def metadata(self) -> Optional[devices.DeviceMetadata]:
+        return self._metadata
+
+    def validate_operation(self, operation: 'cirq.Operation') -> None:
+        for q in operation.qubits:
+            if q not in self._metadata.qubit_set:
+                raise ValueError(f'Qubit not on device: {q!r}.')
+
+        if (
+            len(operation.qubits) == 2
+            and operation.qubits[0] not in self._metadata.nx_graph[operation.qubits[1]]
+        ):
+            raise ValueError(f'Qubit pair is not valid on device: {operation.qubits!r}.')
+
+
+def construct_grid_device(d: int) -> RoutingTestingDevice:
+    qubits = devices.GridQubit.square(d)
+
+    nx_graph = nx.Graph()
+    row_edges = [
+        (devices.GridQubit(i, j), devices.GridQubit(i, j + 1))
+        for i in range(d)
+        for j in range(d - 1)
+    ]
+    col_edges = [
+        (devices.GridQubit(i, j), devices.GridQubit(i + 1, j))
+        for j in range(d)
+        for i in range(d - 1)
+    ]
+    nx_graph.add_edges_from(row_edges)
+    nx_graph.add_edges_from(col_edges)
```

Comment:
people who know more about cirq (@tanujkhattar @verult ): is there not a simpler way of doing this? `GridDevice` maybe? 

I know networkx also has a function for making a 2d rectangular grid. The nodes are `(i, j)`. It's probably easier to make that graph and turn the nodes into `GridQubit(i, j)` than to make the grid qubits and wire them up into a graph

Reply:
Maybe use `TiltedSquareLattice` topology? 

```python
>>> topology = cirq.TiltedSquareLattice(d, d)
>>> nx.relabel_nodes(topology.graph, topology.nodes_to_gridqubits())
```

Reply:
My 2c:
* `GridDevice` wouldn't work for the reason Ammar described. IMO `GridDeviceMetadata` is better for this square device (but very marginally, so no strong preference) because it has the additional `qubit_pairs` property. You'll also get future `GridDeviceMetadata` features for free.
* Using `TiltedSquareLattice` would be harder to read compared to the existing logic IMO because the reader needs to make an extra hop to understand the logic.

But no strong preference - I think all proposed solutions have similar complexity

Reply:
note well that `TiltedSquareLattice` is different than a rectangular 2d grid

Reply:
@tanujkhattar `TiltedSquareLattice` might work to construct rectangular grid graphs but what about ring or butterfly graphs? Then we'd still default to a pure (or networkx) construction, no?